woocommerce/woocommerce-gutenberg-products-block
# Fix 404 and 500 errors in E2E test logs, editor and front-end.

## Motivation

Running the E2E tests and debugging their failures was becoming unwieldy. Our logs were polluted by large amounts of errors, most of which wouldn't make the tests fail, but were in fact captures of the browser's console logs. This made it hard to identify the actual culprits of a failed test run and to parse the logs output.

This PR aims to clean up all those false negatives, and output cleaner logs.

## Changes

- **Change atomic blocks to not use custom webpack build paths**:

Originally, we had treated atomic blocks as different, since they were only supposed to be registered in case a parent component was consuming them. With the advent of Global Styles, and the move towards FSE where the atomic blocks are going to be used as product elements, this distinction made less sense.

Specifically, we had [outputted the built components as custom-named webpack chunks](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2777), which meant that classes inheriting from `AbstractBlock` needed to hardcode this custom path to be compatible.

This meant that we previously had to enqueue scripts in the editor to make sure they were loaded correctly (see, e.g., https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5816), but the class handling of the assets were still trying to fetch them by their default path, causing a slew of `404` errors which weren't really impacting the functionality of the application, but were triggering unnecessary network requests and were caught by Jest during our E2E tests, polluting the logs.

Additionally, on the front-end it also caused more `404` errors, [due to blocks being registered normally as a requirement for enabling Global Styles](https://teamkirigami.wordpress.com/2022/01/28/global-style-what-is-and-how-it-works).

A side-effect of this PR is simplifying this configuration, as now default paths work and don't have to be hard-coded.

This change moves build output for atomic blocks from  `build/atomic-block-components/_name_` to `build/product-_name_` reflecting the actual block name and always prefixed with context namespace (product) to avoid possible name clashes.

- **Ensure all atomic blocks are now registered via php**

Currently AddToCart and SKU blocks were missing as the only blocks that did not support global styles yet, this adds basic setup so that we can unify how 

- **Ensure all lazy loaded blocks (using chunks) register their chunk_translations by themselves via php**

Before AllProducts.php took the responsibility for registering all chunks which was easy to miss or not understand implications. Registered chunks in AllProducts were making them available for usage outside of AllProducts as well.

- **Add missing emptytheme footer.html template part**

The theme `emptytheme` was added as the most minimal block theme possible, in order to cleanly test Site Editor related features. As its purpose was supposed to be getting away with the most minimal of themes, the `footer` template part had not originally been included.

This caused 404 error as this template part is always fetched when user navigates in the site template editor.

- **Replaced trashAllPosts with new REST API based deleteAllTemplates util**

Previously, we used a util called `trashAllPosts` which navigated to the post UI and deleted all the posts to tear down any side-effects of template editing tests.

However, with a [recent change introduced in WordPress 5.9.1](https://github.com/WordPress/wordpress-develop/commit/14e20f72b568cfb5a85f57c77eaa538c17a5f302), Core removed the UI for those and that made our tests meet a 500 error.

Using the REST API should also make all the tests faster.

- **Remove deprecated usage of Puppeteer `waitFor`**

Puppeteer has deprecated the use of the [generic `waitFor`](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pagewaitforselectororfunctionortimeout-options-args) method for more specific methods to avoid confusion.

Deprecation warnings were polluting the logs, so we switched to more specific methods (i.e. mostly changed `waitFor` with `waitForTimeout`). 
  

Fixes #5945
Fixes #5957




### Testing

#### Automated Tests
### Manual Testing

How to test the changes in this Pull Request:

1. Run E2E tests 
2. there should be no 404 or 500 errors in the test logs

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).

1. Visit the editor
2. Add Single Page and All Products blocks 
3. There should be no 404 errors in the editor
4. Visit the created page in the front-end
5. There should be no 404 errors in the page
